src/index.ts

```diff
@@ -88,10 +90,71 @@ class Workflow<T> {
    *
    * NOTE: The uploaded workflow cannot currently be configured here (apart
    * from the name). See the platform UI for configuration of accounts.
+   *
+   * NOTE: UserUtilities that are not functions/classes need to have a name and value.
+   * The value must also be JSON serializable.  They can't refer to external resources
+   * unless they are also included in userUtilities or part of the SDK.
    */
-  async upload() {
+  async upload(options?: { userUtilities?: UserUtility[] }) {
     const fnString = this.fn.toString();
 
+    function mapUserUtilities(s: UserUtility) {
+      if (!(s instanceof Function)) {
+        if (!("value" in s)) {
+          throw new Error("Utility must either be a function/class, have an import_name and function value or have a name and value")
+        }
+        if (s.value instanceof Function) {
+          return s.value.toString();
+        }
+        else {
+          if (!("name" in s)) {
+            throw new Error("Utility must either be a function/class, have an import_name and function value or have a name and value")
+          }
+          return `let ${s.name} = ${JSON.stringify(s.value)}`
+        }
+      }
+      return s.toString()
+    }
+
+    function mapUserUtitlityNames(s: UserUtility): UtilityNaming {
+      var isDefault = false;
+      var import_name = "";
+      var name = "";
+      if ("default" in s) {
+        isDefault = s.default;
+      }
+      if (!(s instanceof Function)) {
+        import_name = "";
+        if ("import_name" in s) {
+          import_name = s.import_name;
+        }
+        else if ("name" in s) {
+          import_name = s.name;
+        }
+
+        if (s.value instanceof Function) {
+          name = s.value.name;
+        }
+        else if ("name" in s) {
+          name = s.name
+        }
+        else {
+          throw new Error("Utility must either be a function/class, have an import_name and function value or have a name and value")
+        }
+
+        import_name = import_name.replace(/[^a-zA-Z0-9_]/g, "_");
+        import_name = import_name.replace(/_+/g, "_");
```

Comment:
Is this the same canonicalisation as the TypeScript compiler? If so then cool, but maybe a comment would be good here.

Reply:
It is partially, but I don't know the whole process of canonicalisation that typescript does and I couldn't find a good resource when I was looking.

Reply:
I have done some additional testing, and is looks like it uses the final part of the path, and this basic canonicalisation is effective so long as there isn't two modules that are the same.  The problem there ends up being _1 and _2 to avoid collisions, and they are numbered based on the order they are first used, not imported.  So I can add that to the documentation and comment these.

Reply:
Added comment and removed the second line